software/template/main.c

```diff
@@ -40,229 +40,254 @@ volatile uint8_t datareg[9]={0};
-        }
-        if(setpoint && temp<setpoint-15){
-            //blink white while heating
-             if(cycles>4){
-                PORTC.OUTSET=2;
-            }else{
-                PORTC.OUTCLR=2;
-            }
-        }
-        if(setpoint && temp>=setpoint-15){
-            PORTC.OUTSET=2;
-        }
-        if(setpoint){
-            pid_harder();
-        }else{
-            duty=0;
-        }
-        if(duty>0){
-        FET_set_level(true);
-        _delay_ms(duty);
+    CLR_PIN(HEATING_LED);
+
+    if (btnignore) {
+      btnignore--;
+      btnflag = 0;
+    } else if (btnflag) {
+      // not ignoring button, first click event:
+      if (setpoint > 0) {
+        // power off on button press, store current setpoint
+        oldsetpoint = setpoint;
+        setpoint    = 0;
+        standby     = 0;
+      } else {
+        // enable power, restore setpoint or use default
+        if (oldsetpoint) {
+          setpoint = oldsetpoint;
+        } else {
+          setpoint = deftemp;
         }
-        FET_set_level(false);
-        datareg[0]=(temp>>1)&0xff;
-        datareg[1]&=~0x1f;
-        datareg[1]=(atemp>>1)&0x1f;
-        datareg[2]=duty;
-        datareg[3]=(setpoint>>1)&0xff;
-        _delay_ms(100-(duty>90?90:duty));
-        cycles++;
-        if(cycles==10){
-            seconds++;
-            if(idlesecs && seconds>=idlesecs){
-                oldsetpoint=setpoint;
-                setpoint=0;
-                standby=1;
-            }
-            cycles=0;
-            if(seconds%60==0){
-                minutes++;
-                if(idlesecs==0){
-                    seconds=0;
-                }else{
-                    if(seconds>idlesecs){
-                        seconds=0;
-                    }
-                }
-                if(offmins){
-                    if(minutes>offmins){
-                        oldsetpoint=setpoint;
-                        setpoint=0;
-                        standby=0;
-                        minutes=0;
-                    }
-                }else{
-                    minutes=0;
-                }
-            }
+        standby = 0;
+      }
+      btnignore = 3;  // ignore button presses for 300ms
+      btnflag   = 0;
+    }
+    if (setpoint && temp < setpoint - 15) {
+      // blink white while heating
+      if (cycles > 4) {
+        SET_PIN(HEATING_LED);
+      } else {
+        CLR_PIN(HEATING_LED);
+      }
+    }
+    if (setpoint && temp >= setpoint - 15) {
+      SET_PIN(HEATING_LED);
+    }
+    if (setpoint) {
+      pid_harder();
+      if(duty > MAX_DUTY_CYCLE) {
+        duty = MAX_DUTY_CYCLE;
+      }
+    } else {
+      duty = 0;
+    }
+    if (duty > 0) {
+      FET_set_level(true);
+      _delay_ms(duty);
+    }
+    FET_set_level(false);
+    datareg[0] = (temp >> 1) & 0xff;
+    datareg[1] &= ~0x1f;
+    datareg[1] = (atemp >> 1) & 0x1f;
+    datareg[2] = duty;
+    datareg[3] = (setpoint >> 1) & 0xff;
+    _delay_ms(100u - duty);
```

Comment:
Can we somehow ensure that duty is never >=100? or is that checked elsewhere?

Reply:
We have `if(duty > MAX_DUTY_CYCLE)` in the main loop, so as long as that's `<= 100` we should be good

Reply:
let's add an explicit check in case someone decides to do horrible things

Reply:
Is the warning in the config.h that I just added enough? I don't really care if people engage in foot-shooting if they've been warned.